File: atividade3.cpp
#include <iostream>
using namespace std;

class PIDController {
private:
    double Kp;
    double Ki;
    double Kd;

    double setpoint = 0.0;
    double measurement = 0.0;
    double integral = 0.0;
    double previous_error = 0.0;

public:
    // Construtor
    PIDController(double Kp_param, double Ki_param, double Kd_param) {
        Kp = Kp_param;
        Ki = Ki_param;
        Kd = Kd_param;
        integral = 0.0;
        previous_error = 0.0;
    }

    // Define o setpoint desejado
    void set_setpoint(double sp) {
        setpoint = sp;
    }


    void update_measurement(double value) {
        measurement = value;
    }


    double compute_output() {
        double error = setpoint - measurement;
        integral += error;                      // erro acumulado
        double derivative = error - previous_error; // derivada
        previous_error = error;

        double output = Kp * error + Ki * integral + Kd * derivative;
        return output;
    }

    
    void display_gains() {
        cout << "Kp: " << Kp
             << " | Ki: " << Ki
             << " | Kd: " << Kd << endl;
    }
};

// Exemplo de uso no main
int main() {
    PIDController pid(1.0, 0.1, 0.05);  
    pid.display_gains();

    double temperatura = 20.0;
    pid.set_setpoint(25.0);             

    for (int i = 0; i < 5; i++) {
        pid.update_measurement(temperatura);

        double sinal = pid.compute_output();
        cout << "Leitura: " << temperatura << " | Sinal de controle: " << sinal << endl;

        
        temperatura += sinal * 0.1; }

File: atividade3.py
class PIDController:
    def __init__(self, Kp: float, Ki: float, Kd: float):
        self._Kp = Kp
        self._Ki = Ki
        self._Kd = Kd

        self._setpoint = 0.0
        self._integral = 0.0
        self._last_error = 0.0

    # Define o setpoint desejado
    def set_setpoint(self, sp: float):
        self._setpoint = sp

    # Atualiza a medição e retorna saída PID
    def update(self, measurement: float) -> float:
        error = self._setpoint - measurement
        self._integral += error
        derivative = error - self._last_error
        self._last_error = error

        output = self._Kp * error + self._Ki * self._integral + self._Kd * derivative
        return output

    # Representação em string mostrando os ganhos
    def __str__(self):
        return f"PID Gains -> Kp: {self._Kp}, Ki: {self._Ki}, Kd: {self._Kd}"


# ---------------------------
# Exemplo de uso / simulação
# ---------------------------
if __name__ == "__main__":
    pid = PIDController(Kp=1.0, Ki=0.1, Kd=0.05)
    print(pid)

    temperatura = 20.0
    pid.set_setpoint(25.0)  # desejamos 25°C

    print("Simulação PID:")
    for i in range(5):
        sinal = pid.update(temperatura)
        print(f"Leitura: {temperatura:.2f} | Sinal de controle: {sinal:.2f}")

        # Simula efeito do controle na temperatura
        temperatura += sinal * 0.1  # fator de ganho fictício
